numpy/lib/function_base.py

```diff
@@ -4014,122 +4192,212 @@ def _quantile_is_valid(q):
+        # After the sort, slices having NaNs will have for last element a NaN
+        virtual_indexes_nans = np.isnan(virtual_indexes)
+        if virtual_indexes_nans.any():
+            previous_indexes[virtual_indexes_nans] = -1
+            next_indexes[virtual_indexes_nans] = -1
+    previous_indexes = previous_indexes.astype(np.intp)
+    next_indexes = next_indexes.astype(np.intp)
+    return previous_indexes, next_indexes
+
+
+def _quantile(
+        arr: np.array,
+        quantiles: np.array,
+        axis: int = -1,
+        interpolation=QuantileInterpolation.INCLUSIVE,
+        out=None,
+):
+    """
+    Private function that doesn't support extended axis or keepdims.
+    These methods are extended to this function using _ureduce
+    See nanpercentile for parameter usage
+    It computes the quantiles of the array for the given axis.
+    A linear interpolation is performed based on the interpolation_type.
+
+    By default, the interpolation is QuantileInterpolation.INCLUSIVE where
+    alpha == beta == 1 which performs the 7th method of Hyndman&Fan.
+    With QuantileInterpolation.MEDIAN_UNBIASED we get  alpha == beta == 1/3
+    thus the 8th method of Hyndman&Fan.
+    """
+    # --- Setup
+    arr = np.asanyarray(arr)
+    values_count = arr.shape[axis]
+    # The dimensions of `q` are prepended to the output shape, so we need the
+    # axis being sampled from `arr` to be last.
+    DATA_AXIS = 0
+    arr = np.moveaxis(arr, axis, destination=DATA_AXIS)
+    # --- Computation of indexes
+    # Index where to find the value in the sorted array.
+    # Virtual because it is a floating point value, not an valid index.
+    # The nearest neighbours are used for interpolation
+    virtual_indexes = interpolation.virtual_index_fun(values_count, quantiles)
+    virtual_indexes = np.asanyarray(virtual_indexes)
+    previous_indexes, next_indexes = _get_indexes(arr,
+                                                  virtual_indexes,
+                                                  values_count)
+    # --- Sorting
+    arr.partition(
+        np.unique(np.concatenate(([0, -1],
+                                  previous_indexes.ravel(),
+                                  next_indexes.ravel(),
+                                  ))),
+        axis=DATA_AXIS)
+    if np.issubdtype(arr.dtype, np.inexact):
+        slices_having_nans = np.isnan(take(arr, indices=-1, axis=DATA_AXIS))
+    else:
+        slices_having_nans = None
+    # --- Get values from indexes
+    previous = np.take(arr, previous_indexes, axis=DATA_AXIS)
+    next = np.take(arr, next_indexes, axis=DATA_AXIS)
+    # --- Linear interpolation
+    gamma = _get_gamma(arr.dtype,
+                       virtual_indexes,
+                       previous_indexes,
+                       interpolation)
+    result_shape = virtual_indexes.shape + (1,) * (arr.ndim - 1)
+    gamma = gamma.reshape(result_shape)
+    result = _linear_interpolation_formula(previous,
```

Comment:
The old code does not call the interpolation formula when the dtype of the indices is integer I think (i.e. gamma is unimportant?)

In particular:
```
np.quantile(np.array([1, 2], dtype=np.uint8), [0.5], interpolation="nearest")
```
still returns integers (because you do cast to `intp` in the `gamma` formula) but not `uint8` anymore.  (Or am I missing the bool regression thing in current NumPy that was rolled back in main, but not here?)

Reply:
Yes previously when dtype of indices was integer it was short-cutting the interpolation because with an integer we can directly pickup the element in the array whichever is the interpolation method.
However it's writing a specific case when the general case seems to work fine which I thought was the case here. With integer indices, `gamma` is zero thus the interpolation would gives the left element.

But maybe it was made to handle some cases such as `uint8`.
Especially, I didn't pay that much attention to lower, higher, nearest which should preserve dtype.

Should I re-add this Integer handling ?

Reply:
Please do, I think it makes more sense for the discrete methods in general to be honest.